src/Chapter3.hs

```diff
@@ -479,7 +510,54 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
+data Castle
+    = Castle String
+    | WalledCastle String
+    | None
+    deriving (Eq, Show)
+
+data ChurchLib
+    = Church
+    | Library
+    deriving (Show)
+
+data House
+    = One
+    | Two
+    | Three
+    | Four
+    deriving (Eq, Show)
+
+data City = City
+    { cityCastle :: Castle
+    , cityChurchLib :: ChurchLib
+    , cityHouses :: [House]
+    } deriving (Show)
 
+buildCastle :: City -> String -> City
+buildCastle city@City {cityCastle = WalledCastle _} name =
+    city {cityCastle = WalledCastle name}
+buildCastle city name =
+    city {cityCastle = Castle name}
+
+buildHouse :: City -> House -> City
+buildHouse city house =
+    city {cityHouses = cityHouses city ++ [house]}
+
+convert :: House -> Int
+convert One = 1
+convert Two = 2
+convert Three = 3
+convert _ = 4
```

Comment:
For the sum types it's usually recommended to handle all constructors explicitly instead of using wildcard. Let's say that tomorrow you need to add the constructor `Five` to the `House` data type. If you write all constructors explicitly, the compiler warns you when you forget to pattern match, so it becomes much easier to refactor and add new features :slightly_smiling_face: 

Also, here, to avoid some code duplication, you can use the `case-of` expression:

```haskell
convert house = case house of
    One -> 1
    ... -> ...
```